widgets/viewports/renderer.cpp

```diff
@@ -1903,6 +1890,129 @@ hash_list<nodeListElement *> ViewportBase::pickNodes(int centerX, int centerY, i
+
+        state->viewerState->colorPickingBuffer24.emplace_back(getPickingColor(node, RenderOptions::SelectionPass::NodeID0_24Bits));
+        state->viewerState->colorPickingBuffer48.emplace_back(getPickingColor(node, RenderOptions::SelectionPass::NodeID24_48Bits));
+        state->viewerState->colorPickingBuffer64.emplace_back(getPickingColor(node, RenderOptions::SelectionPass::NodeID48_64Bits));
+        state->viewerState->pointVertBuffer.emplace_back(isoPos, arrayFromQColor(color));
+    };
+
+    const auto * activeTree = state->skeletonState->activeTree;
+    const auto * activeNode = state->skeletonState->activeNode;
+    const auto * activeSynapse = (activeNode && activeNode->isSynapticNode) ? activeNode->correspondingSynapse :
+                                 (activeTree && activeTree->isSynapticCleft) ? activeTree->correspondingSynapse :
+                                                                               nullptr;
+    const bool synapseBuilding = state->skeletonState->synapseState != Synapse::State::PreSynapse;
+    const bool onlySelected = state->viewerState->skeletonDisplay.testFlag(TreeDisplay::OnlySelected);
+
+    for (auto & currentTree : Skeletonizer::singleton().skeletonState.trees) {
+        // focus on synapses, darken rest of skeleton
+        const bool darken = (synapseBuilding && currentTree.correspondingSynapse != &state->skeletonState->temporarySynapse)
+                || (activeSynapse && activeSynapse->getCleft() != &currentTree);
+        const bool hideSynapses = !darken && !synapseBuilding && !activeSynapse && currentTree.isSynapticCleft;
+        const bool selectionFilter = onlySelected && !currentTree.selected;
+        if (selectionFilter || (!currentTree.render && (!currentTree.isSynapticCleft || (currentTree.isSynapticCleft && hideSynapses)) )) {
+            // hide synapse takes precedence over render flag for synapses.
+            continue;
+        }
+        nodeListElement * previousNode = nullptr;
+
+        for (auto nodeIt = std::begin(currentTree.nodes); nodeIt != std::end(currentTree.nodes); ++nodeIt) {
+
+            //This sets the current color for the segment rendering
+            QColor currentColor = currentTree.color;
+            if (state->viewerState->highlightActiveTree && currentTree.treeID == state->skeletonState->activeTree->treeID) {
+                currentColor = Qt::red;
+            }
+            if (darken) {
+                currentColor.setAlpha(Synapse::darkenedAlpha);
+            }
+
+            for (const auto & currentSegment : nodeIt->segments) {
+                if (currentSegment.forward) {
+                    continue;
+                }
+                addSegment(currentSegment, currentColor);
+            }
+
+            addNode(*nodeIt);
+
+            previousNode = &*nodeIt;
+        }
+    }
+
+    /* Connect all synapses */
+    for (auto & synapse : state->skeletonState->synapses) {
```

Comment:
i don’t know if this could be better short of putting the whole loop into a function which takes the different `addSegment`s as argument